File: client/golang/types/workflow.go
/*
 * Copyright (c) 2000, 2023, ducesoft and/or its affiliates. All rights reserved.
 * DUCESOFT PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */

package types

var (
	WorkStart  VertexKind = 1
	WorkFinish VertexKind = 2
	WorkJob    VertexKind = 4
	WorkMan    VertexKind = 8
	WorkWay    VertexKind = 16
	WorkTimer  VertexKind = 32
)

type VertexKind int64

type WorkIntent struct {
	BNO     string            `index:"0" json:"bno" yaml:"bno" xml:"bno" comment:"Business code"`
	CNO     string            `index:"1" json:"cno" yaml:"cno" xml:"cno" comment:"Workflow chart code"`
	Context map[string]string `index:"2" json:"context" yaml:"context" xml:"context" comment:"Workflow context"`
	Applier *Worker           `index:"3" json:"applier" yaml:"applier" xml:"applier" comment:"Workflow applier"`
}

type WorkVertex struct {
	Name  string            `index:"0" json:"name" yaml:"name" xml:"name" comment:"Workflow name"`
	Alias string            `index:"1" json:"alias" yaml:"alias" xml:"alias" comment:"Workflow alias"`
	Attrs map[string]string `index:"2" json:"attrs" yaml:"attrs" xml:"attrs" comment:"Workflow vertex attributes"`
	Kind  int64             `index:"3" json:"kind" yaml:"kind" xml:"kind" comment:"Workflow vertex kind"`
	Group string            `index:"4" json:"group" yaml:"group" xml:"group" comment:"Workflow review group code"`
}

type WorkSide struct {
	Src       string `index:"0" json:"src" yaml:"src" xml:"src" comment:"Workflow side src name"`
	Dst       string `index:"1" json:"dst" yaml:"dst" xml:"dst" comment:"Workflow side dst name"`
	Condition string `index:"2" json:"condition" yaml:"condition" xml:"condition" comment:"Workflow side condition"`
}

type WorkChart struct {
	CNO        string        `index:"0" json:"cno" yaml:"cno" xml:"cno" comment:"Workflow chart code"`
	Name       string        `index:"1" json:"name" yaml:"name" xml:"name" comment:"Workflow name"`
	Vertices   []*WorkVertex `index:"2" json:"vertices" yaml:"vertices" xml:"vertices" comment:"Workflow vertices"`
	Sides      []*WorkSide   `index:"3" json:"sides" yaml:"sides" xml:"sides" comment:"Workflow sides"`
	Status     int64         `index:"4" json:"status" yaml:"status" xml:"status" comment:"Workflow status"`
	Maintainer *Worker       `index:"5" json:"maintainer" yaml:"maintainer" xml:"maintainer" comment:"Workflow maintainer"`
}

func (that *WorkChart) Vertex(vertex *WorkVertex) {
	that.Vertices = append(that.Vertices, vertex)
}

func (that *WorkChart) Link(side *WorkSide) {
	that.Sides = append(that.Sides, side)
}

type WorkRoutine struct {
	RNO     string            `index:"0" json:"rno" yaml:"rno" xml:"rno" comment:"Workflow routine code"`
	BNO     string            `index:"1" json:"bno" yaml:"bno" xml:"bno" comment:"Business code"`
	Context map[string]string `index:"2" json:"context" yaml:"context" xml:"context" comment:"Workflow context"`
	Status  int64             `index:"3" json:"status" yaml:"status" xml:"status" comment:"Workflow status"`
	Chart   *WorkChart        `index:"4" json:"chart" yaml:"chart" xml:"chart" comment:"Workflow chart"`
	Tasks   []*WorkTask       `index:"5" json:"tasks" yaml:"tasks" xml:"tasks" comment:"Workflow tasks"`
}

type WorkTask struct {
	Vertex    *WorkVertex       `index:"0" json:"vertex" yaml:"vertex" xml:"vertex" comment:"Workflow vertex"`
	Reviewers []*Worker         `index:"1" json:"reviewers" yaml:"reviewers" xml:"reviewers" comment:"Workflow vertex reviewers"`
	Status    int64             `index:"2" json:"status" yaml:"status" xml:"status" comment:"Workflow vertex status"`
	Context   map[string]string `index:"3" json:"context" yaml:"context" xml:"context" comment:"Workflow context"`
}

type WorkGroup struct {
	NO      string    `index:"0" json:"no" yaml:"no" xml:"no" comment:"Work group identity"`
	Name    string    `index:"1" json:"name" yaml:"name" xml:"name" comment:"Work group name"`
	Status  int64     `index:"2" json:"status" yaml:"status" xml:"status" comment:"Workflow group status"`
	Workers []*Worker `index:"3" json:"workers" yaml:"workers" xml:"workers" comment:"Work group workers"`
}

type Worker struct {
	NO    string `index:"0" json:"no" yaml:"no" xml:"no" comment:"Worker identity"`
	Name  string `index:"1" json:"name" yaml:"name" xml:"name" comment:"Worker name"`
	Alias string `index:"2" json:"alias" yaml:"alias" xml:"alias" comment:"Worker alias"`
}

type WorkContext struct {
	RNO      string            `index:"0" json:"rno" yaml:"rno" xml:"rno" comment:"Workflow routine code"`
	BNO      string            `index:"1" json:"bno" yaml:"bno" xml:"bno" comment:"Business code"`
	CNO      string            `index:"2" json:"cno" yaml:"cno" xml:"cno" comment:"Workflow chart code"`
	Context  map[string]string `index:"3" json:"context" yaml:"context" xml:"context" comment:"Workflow context"`
	Vertex   *WorkVertex       `index:"4" json:"vertex" yaml:"vertex" xml:"vertex" comment:"Workflow vertex"`
	Task     WorkTask          `index:"5" json:"task" yaml:"task" xml:"task" comment:"Workflow task"`
	Applier  *Worker           `index:"6" json:"applier" yaml:"applier" xml:"applier" comment:"Workflow applier"`
	Reviewer *Worker           `index:"7" json:"reviewer" yaml:"reviewer" xml:"reviewer" comment:"Workflow reviewer"`
}

File: client/javascript/src/kinds/workflow.ts
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */

import {index} from "@/macro";

export enum VertexKind {
    WorkStart = 1,
    WorkFinish = 2,
    WorkJob = 4,
    WorkMan = 8,
    WorkWay = 16,
    WorkTimer = 32,
}

export class Worker {
    /**
     * Workflow worker code
     */
    @index(0)
    public no: string = "";
    /**
     * Workflow worker name
     */
    @index(1)
    public name: string = "";
    /**
     * Workflow worker alias
     */
    @index(2)
    public alias: string = "";
}

export class WorkIntent {
    /**
     * Business code
     */
    @index(0)
    public bno: string = "";
    /**
     * Workflow chart code
     */
    @index(1)
    public cno: string = "";
    /**
     * Workflow context
     */
    @index(2)
    public context: Map<string, string> = new Map();
    /**
     * Workflow applier
     */
    @index(3)
    public applier: Worker = new Worker();
}

export class WorkVertex {
    /**
     * Workflow name
     */
    @index(0)
    public name: string = "";
    /**
     * Workflow alias
     */
    @index(1)
    public alias: string = "";
    /**
     * Workflow context
     */
    @index(2)
    public attrs: Map<string, string> = new Map();
    /**
     * Workflow kind
     */
    @index(3)
    public kind: number = 0;
    /**
     * Workflow review group
     */
    @index(4)
    public group: string = "";
}

export class WorkSide {
    /**
     * Workflow side src name
     */
    @index(0)
    public src: string = "";
    /**
     * Workflow side dst name
     */
    @index(1)
    public dst: string = "";
    /**
     * Workflow side condition
     */
    @index(2)
    public condition: string = "";
}

export class WorkChart {
    /**
     * Workflow chart code
     */
    @index(0)
    public cno: string = "";
    /**
     * Workflow chart name
     */
    @index(1)
    public name: string = "";
    /**
     * Workflow vertices
     */
    @index(2)
    public vertices: WorkVertex[] = [];
    /**
     * Workflow sides
     */
    @index(3)
    public sides: WorkSide[] = [];
    /**
     * Workflow status
     */
    @index(3)
    public status: number = 0;
    /**
     * Workflow maintainer
     */
    @index(4)
    public maintainer: Worker = new Worker();

    public vertex(vertex: WorkVertex): void {
        this.vertices.push(vertex);
    }

    public link(side: WorkSide): void {
        this.sides.push(side);
    }

}


export class WorkRoutine {
    /**
     * Workflow routine code
     */
    @index(0)
    public rno: string = "";
    /**
     * Workflow business code
     */
    @index(1)
    public bno: string = "";
    /**
     * Workflow context
     */
    @index(2)
    public context: Map<string, string> = new Map();
    /**
     * Workflow status
     */
    @index(3)
    public status: number = 0;
    /**
     * Workflow chart
     */
    @index(4)
    public chart: WorkChart = new WorkChart();
    /**
     * Workflow tasks
     */
    @index(5)
    public tasks: WorkChart[] = [];
}

export class WorkTask {
    /**
     * Workflow vertex
     */
    @index(0)
    public vertex: WorkVertex = new WorkVertex();
    /**
     * Workflow reviewers
     */
    @index(1)
    public reviewers: Worker[] = [];
    /**
     * Workflow status
     */
    @index(2)
    public status: number = 0;
    /**
     * Workflow context
     */
    @index(3)
    public context: Map<string, string> = new Map();
}

export class WorkGroup {
    /**
     * Workflow group code
     */
    @index(0)
    public no: string = "";
    /**
     * Workflow group name
     */
    @index(1)
    public name: string = "";
    /**
     * Workflow status
     */
    @index(2)
    public status: number = 0;
    /**
     * Workflow reviewers
     */
    @index(3)
    public reviewers: Worker[] = [];
}

export class WorkContext {
    /**
     * Workflow routine code
     */
    @index(0)
    public rno: string = "";
    /**
     * Workflow business code
     */
    @index(1)
    public bno: string = "";
    /**
     * Workflow chart code
     */
    @index(2)
    public cno: string = "";
    /**
     * Workflow context
     */
    @index(3)
    public context: Map<string, string> = new Map();
    /**
     * Workflow vertex
     */
    @index(4)
    public vertex: WorkVertex = new WorkVertex();
    /**
     * Workflow task
     */
    @index(5)
    public task: WorkTask = new WorkTask();
    /**
     * Workflow applier
     */
    @index(6)
    public applier: Worker = new Worker();
    /**
     * Workflow reviewer
     */
    @index(7)
    public reviewer: Worker = new Worker();
}
